Add waterfurnace platform

sdague opened the pull request:
This adds support for waterfurnace geothermal systems. This is
implemented as a component as there will eventually be some active
control elements. This is not done as a climate platform because
geothermal systems work best when set at a constant temperature as
they are tuned to keep within 0.5 degrees F of a setpoint, and large
temperature shifts are slow and expensive.

This is done in the Data + Sensors model, with the Data component
having a regular update thread. That thread needs to call the read()
function at least every 30 seconds otherwise the underlying websocket
is closed by the server.

## Description:

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml
waterfurnace:
  username: <myuser>
  password: <mypasswd>
  unit: <unitserialnumber>
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

sdague committed:
Add waterfurnace platform

This adds support for waterfurnace geothermal systems. This is
implemented as a component as there will eventually be some active
control elements. This is not done as a climate platform because
geothermal systems work best when set at a constant temperature as
they are tuned to keep within 0.5 degrees F of a setpoint, and large
temperature shifts are slow and expensive.

This is done in the Data + Sensors model, with the Data component
having a regular update thread. That thread needs to call the read()
function at least every 30 seconds otherwise the underlying websocket
is closed by the server.

MartinHjelmare commented on homeassistant/components/sensor/waterfurnace.py:
```diff
@@ -0,0 +1,114 @@
+            self.async_schedule_update_ha_state()
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Please move this above the entity class.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/waterfurnace.py:
```diff
@@ -0,0 +1,91 @@
+import logging
+import time
+import threading
+import voluptuous as vol
```
Add a blank line between import groups standard library, 3rd party and homeassistant.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/waterfurnace.py:
```diff
@@ -0,0 +1,91 @@
+    unit = config.get(CONF_UNIT)
+
+    wfconn = wf.WaterFurnace(username, password, unit)
+    wfconn.login()
```
What happens if login fails?

sdague replied:
Right now, a KeyError is thrown because sessionid is not found in the cookies. It's not elegant, but it will stop the rest of the execution.

MartinHjelmare commented on homeassistant/components/waterfurnace.py:
```diff
@@ -0,0 +1,91 @@
+                self.client.login()
+                _LOGGER.error("Lost our connection to websocket, trying again")
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Error updating waterfurnace data.")
```
If the credentials are bad, this will just keep trying. That's not efficient.

sdague replied:
The login happening inside the except handler will make trigger a second exception which will stop it. It's maybe not the most elegant, but it shouldn't loop forever.

MartinHjelmare replied:
You're correct. For some reason I thought the second handler was nested. Can you add an exception handler for the key error that logs an error and breaks the loop?

sdague replied:
ack

MartinHjelmare reviewed (changes requested):
Looks good overall. You should clean up the thread when home assistant fires the home assistant stop event.

See below for more comments.

sdague committed:
Fix for review comments

MartinHjelmare commented on homeassistant/components/waterfurnace.py:
```diff
@@ -0,0 +1,117 @@
+    wfconn = wf.WaterFurnace(username, password, unit)
+    # NOTE(sdague): login will throw an exception if this doesn't
+    # work, which will abort the setup.
+    wfconn.login()
```
Please add a try except block around this and return `False` upon error.

sdague replied:
ack

MartinHjelmare reviewed (changes requested):
See comment below. Please also do something similar in the thread loop.

sdague committed:
Address more review feedback

This includes being more careful about exceptions when login() might
fail. A new version of the waterfurnace library is needed as that
created library specific exceptions to catch.

sdague commented:
I think I've got all the feedback addressed, take a look when you can.

MartinHjelmare reviewed (approved):
Looks good!

An improvement would be to catch more specific errors from the library in the thread loop. But I think it's ok like this.

MartinHjelmare commented:
Link a docs PR in the description and this can be merged.

sdague commented:
Merging given that the docs are up and the ACK was done after the last set of revisions
